### ISSUE TITLE
Make it easier to access engine API

### ISSUE THREAD

**Issue by errordeveloper** [docker/docker#27261]
I suppose would be rather very fine if there was a way to say "This container need to be able to call Docker client" with just one flag for `docker run` (`--with-engine-api-access`?) and have client library or command that would work with older and newever versions of the daemon (same major version?).

**Comment by errordeveloper** [comment]
> ...and have client library or command that would work with older and newever versions of the daemon (same major version?)

Seems like https://github.com/docker/docker/issues/25498 should cover this.

**Comment by justincormack** [comment]
Sorry, can you explain in more detail? You want Docker socket access? We looked at some of the issues in https://github.com/docker/docker/pull/25347 but have postponed for now as there were many issues, but it is planned.

**Comment by errordeveloper** [comment]
> You want Docker socket access?

Yes, with a self-explanatory flag and without having to care about the path on either side.

I've read the discussion in #25347 briefly just now. I think it shouldn't be done pure with `--privileged` flag, a separate flag would be needed, I'd imagine...

**Comment by justincormack** [comment]
The reason for initially limiting it to --privileged was that it does give
you complete access, including to privileged on the host. Doing anything
else requires comprehensive filtering.

As there are a lot of use cases filtering gets complicated. So more
specific use cases need to be considered.

**Comment by AkihiroSuda** [comment]
My usecase would be a cron-like service that executes `docker service create --restart-condition none some-batch-job` periodically (https://github.com/docker/docker/issues/23880#issuecomment-237343613, cc @stevvooe )

For filtering, I'd like to suggest this model:

Name (FIXME) | Allows | Prohibits | Use case
---|---|---|---
`DOCKER_CAP_FORK` |  <ul><li>creating a new container and listing/inspecting/removing it. <li> `docker images`<li>`docker build`<li>`docker pull`<li>`docker events` (only about newly-created containers) <li>`docker network create`? <li>`docker volume create`?</ul> |<ul><li>listing/inspecting/removing existing containers. <li>`docker login`? <li>`docker plugin install`<li>`docker node`<li>`docker swarm`</ul> | cron-like service (please see above)
`DOCKER_CAP_INSPECT` | read-only operation such as listing and inspecting existing containers. Note that `docker inspect` can leak a secret environment variable (if set) | read-write operation such as creating or removing a container | Monitoring service
`DOCKER_CAP_ADMIN` | Everything | None | unsure :confused:

**Comment by justincormack** [comment]
@AkihiroSuda yes, I looked at a similar split with @bfirsh a while back.

**Comment by archisgore** [comment]
I want to add to this that we'd love to get the Swarm API endpoint as well. Engine API is useful, but as we started migrating to swarms, we have a hard time creating the self-referential links to the swarm.

If whitelisted services were able to provide a uniform abstracted API into the outer swarm, it would greatly make life easier for us.
